modules/parquet/src/parquetjs/reader.ts

```diff
@@ -466,46 +496,74 @@ async function decodeDataPages(
     rlevels: [],
     dlevels: [],
     values: [],
+    pageHeaders: [],
     count: 0
   };
 
-  // @ts-ignore size can be undefined
-  while (cursor.offset < cursor.size) {
+  while (
+    // @ts-ignore size can be undefined
+    cursor.offset < cursor.size &&
+    (!opts.num_values || data.dlevels.length < opts.num_values)
+  ) {
     // const pageHeader = new parquet_thrift.PageHeader();
     // cursor.offset += parquet_util.decodeThrift(pageHeader, cursor.buffer);
 
-    const {pageHeader, length} = decodePageHeader(cursor.buffer);
-    cursor.offset += length;
+    const page = decodePage(cursor, opts);
+
+    if (page.dictionary) {
+      opts.dictionary = page.dictionary;
```

Comment:
It is preferable to not modify input parameters (opts in the case), as this can be surprising to the calling function.

Reply:
Changed logic here.